### ISSUE TITLE
GridView columns shortcut declarations: allow value function as value 

### ISSUE THREAD

**Issue by robsch** [yiisoft/yii2#16624]
It would be nice if the columns could be defined this way:

```php
<?= GridView::widget([
    'dataProvider' => $dataProvider,
    'filterModel'  => $searchModel,
    'columns'      => [
        // ---- current style ----
        [
            'attribute' => 'full_name',
            'label'     => 'Name',
            'format'    => 'html',
            'value'     => function (Order $model) {
                return Html::tag('span', $model->first_name . '<br>' . $model->last_name);
            },
        ],
        // ---- new alternative style ----
        'full_name:html:Name' => function(Order $model) {
            return Html::tag('span', $model->first_name . '<br>' . $model->last_name);
        },
        ...
        ['class' => 'yii\grid\ActionColumn'],
    ],
]); ?>
```
Would that be possible? Because of formatting I need to declare columns explicitly quite often. This would reduce the lines of code a bit.

**Comment by rob006** [comment]
I would prefer some helper which will provide API similar to defining form fields:

```php
Grid::column('full_name')->asHtml()->label('Name')->value(function(Order $model) {
    return Html::tag('span', $model->first_name . '<br>' . $model->last_name);
}),
```

**Comment by robsch** [comment]
Yes, that is also nice!

**Comment by codespede** [comment]
> I would prefer some helper which will provide API similar to defining form fields:
> 
> ```
> Grid::column('full_name')->asHtml()->label('Name')->value(function(Order $model) {
>     return Html::tag('span', $model->first_name . '<br>' . $model->last_name);
> }),
> ```
The property `\yii\grid\Column::grid` is mandatory for the columns to render. Hence the possibility is, we should accept Closures for single column definitions in which we can make the instantiated GridView object available which in turn facilitates the proposal of @rob006 as below:

```
GridView::widget([
    'dataProvider' => $dataProvider
    'columns' => [
    	'title',
    	function($grid){
            return GridView::column('description', $grid)->asHtml()->label('Name')->value(function(Order $model) {
                return Html::tag('span', $model->first_name . '<br>' . $model->last_name);
            });
         },
      ],
])
```

**Comment by robsch** [comment]
All interesting and reasonable. But don't forget the _shortcut declaration_. I find it quite handy...

**Comment by rob006** [comment]
> The property \yii\grid\Column::grid is mandatory for the columns to render

The idea is to generate array or entity with column configuration. You don't need to create `Column` instance here, it could be initialized inside of `GridView`.

**Comment by codespede** [comment]
Agree. I'm ready to make a PR.

**Comment by SamMousa** [comment]
I don't think this should be in a separate class though. It's essentially like `QueryBuilder`.
So in my opinion it would be non static syntax like this:
```php
(new GridColumnBuilder('full_name'))
    ->asHtml()
    ->label('test 13')
    ->value(function($model) { return $model->prop; })
    ->create()
;
```

Don't have an opinion on the class name. 
Note that this class should not have any dependencies, all it should do is create the array and return.

**Comment by cebe** [comment]
This sounds like making it all more complicated, creating a builder class for creating a config for configuring an object? This adds a new different syntax for the same thing that is currently already possible but makes it more complex.

**Comment by SamMousa** [comment]
@cebe, longer is not more complex. Code completion makes this simpler than memorizing array keys...

**Comment by samdark** [comment]
That's rather controversial. Builder should be fine but in the variant provided above it's indeed, more complicated than current config.

**Comment by samdark** [comment]
I guess we need proof of concept...

**Comment by SamMousa** [comment]
Note that what I currently do for similar cases is create a builder that creates the array config the widget uses. That keeps it light, I'll see if I have time for this, lower priority for me than Yii3

**Comment by rob006** [comment]
I'm already using helpers for generating grid columns in Yii 1.1 app. It's like

```php
GridColumn::boolean('active'),
```

vs 

```php
[
	'name' => 'active',
	'type' => 'boolean',
	'filter' => [
		1 => Yii::t('admin', 'Yes'),
		0 => Yii::t('admin', 'No'),
	],
],
```

And you can win even more with custom formats or column types.

Helpers are just more flexible version of current `name:type:label` string syntax. And you can always override/add some settings by additional method, without need to completely change existing syntax.
